openmdao/solvers/solver_base.py

```diff
@@ -1,9 +1,70 @@
 """ Base class for linear and nonlinear solvers."""
 
 from __future__ import print_function
+
+from functools import wraps
+import sys
+from six import iterkeys, PY3, reraise
+
+import numpy as np
+
 from openmdao.recorders.recording_manager import RecordingManager
 from openmdao.util.options import OptionsDictionary
 
+def error_wrap_nl(fn):
+    """ Decorator adds some error-handling for floating point errors to any
+    driver function."""
+
+    @wraps(fn)
+    def wrapper(driver, params, unknowns, resids, system, metadata):
+        """ Mainly so sphinx autodoc works"""
+
+        try:
+            fn(driver, params, unknowns, resids, system, metadata)
+        except FloatingPointError as err:
+            exc_info = sys.exc_info()
+
+            # So we don't keep re-appending in a solver stack.
+            if hasattr(exc_info[1], 'seen'):
+                reraise(exc_info[0], exc_info[1], exc_info[2])
+
+            # The user may need some help figuring things out, so let them know where
+            x_unknowns = []
+            for var in iterkeys(unknowns):
+                if unknowns.metadata(var).get('pass_by_obj'):
+                    continue
+                if not all(np.isfinite(unknowns._dat[var].val)):
+                    x_unknowns.append(var)
+            x_resids = []
+            for var in iterkeys(resids):
+                if resids.metadata(var).get('pass_by_obj'):
+                    continue
+                if not all(np.isfinite(resids._dat[var].val)):
+                    x_resids.append(var)
+            x_params = []
+            for var in iterkeys(params):
+                if params.metadata(var).get('pass_by_obj'):
+                    continue
+                if not all(np.isfinite(params._dat[var].val)):
+                    x_params.append(var)
+
+            msg = exc_info[1].message
```

Comment:
Looks like https://docs.python.org/2.7/library/traceback.html#traceback.format_exc is the compatible way to grab this.